feat: Support for using Ansible playbooks for installing model components (#113)
sandialabs/firewheel: docs/source/model_component/mc_install.rst

*************************
INSTALL File Requirements
*************************

Regardless of whether the INSTALL file is an Ansible Playbook or an executable, FIREWHEEL expects the following:

1. Upon successful installation, a new file is created in the model component directory with the following format: ``.<MC Name>.installed``. For example, if the model component name is ``dns.dns_objects`` than the new file would be ``.dns.dns_objects.installed``.
2. After execution, all non-packaged dependencies for the model component should be met.

*****************
Design Principles
*****************

In addition to the requirements above, we recommend that the following principles are adhered to when creating a new INSTALL file.

1. `Idempotence <https://en.wikipedia.org/wiki/Idempotence>`_ - The file should be capable of running multiple times without causing issues. This is a core tenant of Ansible and a strong motivator why Ansible Playbooks are the preferred INSTALL file method.
2. **Reproducibility** - It is critical that users will download the exact same data that was originally intended by the Model Component creators.

>    If the data/packages differ, than there is a strong possibly that the experimental outcomes will differ and could produce unintended consequences.

Reviewer: possibly -> possibility

6. **Readability** - Users will need to execute these potentially unknown actions, the INSTALL script should be well documented and readable to the average user. Readability is desired over brevity.

.. _mc_install_ansible:

*********************
Ansible INSTALL Files
*********************

While INSTALL files should not have an extension, if INSTALL contains valid YAML, is a list, and has the "hosts" key in each list entry, then FIREWHEEL will attempt to use Ansible to execute the file.
We strongly recommend using ``localhost`` as the "hosts" value.
We also recommend that users define an ``install_flag`` variable and a ``cached_files`` variable (if needed).

Below is beginning of an example Ansible "play":

.. code-block:: yaml

    ---
    - name: Download DNS Data files
      hosts: localhost
      become: yes
      vars:
        install_flag: "{{ playbook_dir }}/.dns.dns_objects.installed"
        cached_files:
          - source: "firewheel_repo_dns/dns_objects/bind9_xenial_debs.tgz"
            destination: "vm_resources/bind9_xenial_debs.tgz"

> When designing the tasks, recall that it is possible that the existing ``cached_files`` to become available prior to playbook execution.

Reviewer: I have no idea what this line is saying

************
Cached Files
************

FIREWHEEL supports collecting pre-computed blobs from various resources to enable offline experiment access.

> To enable retrieving files from a cache, users should set the ``ansible.cache_type`` to ``git`` or ``s3`` depending on if the files are cached in a git repository or in an Amazon S3 data store.

Reviewer: Still working through the entire architecture of this, so forgive is this is wrong -- but doesn't this force all model components to use the same cache_type? That feels pretty limiting, but maybe there's a way around it that I'm not seeing yet

Author: That is sort of correct and this discussion is exactly why I left this PR in draft mode :). Generally people can call the INSTALL scripts in two different ways. They can execute the model component (via `firewheel experiment`) which will check if the MC needs to be installed. Alternatively, users can run `firewheel repository install --script` which checks for `INSTALL` files in all repositories and runs them.

You are correct that each set of model components that is installed at once must use the same `cache_type`. However, it is feasible modify the `cache_type` between installing another set of components. So it is rather inflexible for a given set of components, but when installing a new repo, can be modified.

The primary (possibly incorrect) assumption is that organizations that wish to cache files will likely use the same methodology for all MC repos. If an organization uses multiple cache types that would likely become a maintenance burden at some point. However, if different groups within an organization were responsible for managing different MC repos, than it seems very feasible that multiple cache types are needed. In this case, the thought was that users would install one set and then install the second set, modifying the `cache_type` in between.

It seems reasonable to address this limitation, but it would likely add some complexity to the system. Without too much prior thought a new CLI argument for `repository install` may be used to override this variable "on the fly". However, that wouldn't address `firewheel experiment`.  Thoughts and discussion are much appreciated!

    firewheel_repo_dns -- Cloned repository
    └── firewheel_repo_dns
        └── dns_objects
            └── bind9_xenial_debs.tgz

URL Cache
=========
If users have access to any file server (HTTP/HTTPS/FTP), they can specify the ``ansible.cache_type`` key as ``url`` in the :ref:`firewheel_configuration`.
Additional configuration options under the ``ansible`` key are also necessary.
If these values are not provided, but ``ansible.cache_type`` is ``url``, the user will be prompted for the information.

- ``url`` - The URL of the server hosting the cached files.
- ``url_cache_path`` - The path to the file's base directory from the server. For example: ``http://example.com/<url_cache_path>/file.txt``.
- ``use_proxy`` - (optional) If ``false``, it will not use a proxy, even if one is defined in an environment variable on the target hosts. The default is: ``true``.
- ``validate_certs`` - (optional) If ``false``, SSL certificates will not be validated. The default is: ``true``.

If you are using an username or password token, you can specify it in the URL.
For example: ``https://user:password@server.com/url/cache/path/file.txt``

Git Cache
=========

> If users have access to an git server instance, they can specify the ``ansible.cache_type`` key as ``git`` in the :ref:`firewheel_configuration`.

Reviewer: an -> a

Additional configuration options under the ``ansible`` key are also necessary.
If these values are not provided, but ``ansible.cache_type`` is ``git``, the user will be prompted for the information.

- ``git_server`` - The URL of the git server.
- ``git_repo_path`` - The path to the repo from the server. Because this is likely to change for each model component, we recommend not setting this parameter and simply prompting the user for each path.
- ``git_branch`` - (optional) The branch name, defaults to ``main``.

> If you an access token is being used, the user can specify it in the ``git_server`` URL.

Reviewer: get rid of "you"

If these values are not provided, but ``ansible.cache_type`` is ``url``, the user will be prompted for the information.

- ``url`` -- The URL of the server hosting the cached files.
- ``url_cache_path`` -- The path to the file's base directory from the server. For example: ``http://example.com/<url_cache_path>/file/path/file.txt``.
- ``use_proxy`` -- (optional) If ``false``, it will not use a proxy, even if one is defined in an environment variable on the target hosts. The default is: ``true``.
- ``validate_certs`` -- (optional) If ``false``, SSL certificates will not be validated. The default is: ``true``.

If you are using an username or password token, you can specify it in the URL.
For example: ``https://user:password@server.com/url/cache/path/file/path/file.txt``

Git Cache
=========
If users have access to a git server instance, they can specify the ``ansible.cache_type`` key as ``git`` in the :ref:`firewheel_configuration`.
Additional configuration options under the ``ansible`` key are also necessary.
If these values are not provided, but ``ansible.cache_type`` is ``git``, the user will be prompted for the information.

- ``git_server`` -- The URL of the git server.

> - ``git_repo_path`` -- The path to the repo from the server. Because this is likely to change for each model component, we recommend not setting this parameter and simply prompting the user for each path.

Reviewer: the `git_repo_path` parameter should really go in the INSTALL file. Relying on a user specifying the git repo is very non-conducive to automated installation

Author: Maybe this is partly a variable name issue. ``git_repo_path`` is not the exact git repo, but rather the group name + the repo name. For this repo here is what it would look like:

- `git_server=https://github.com`
- `git_repo_path=sandialabs/firewheel.git`

I suspect that maybe this should be in three different variables. It doesn't make sense for the `INSTALL` files to specify the group name, but maybe forcing the same "repo name" makes sense?

- `git_server=https://github.com`
- `git_group_path=sandialabs`
- `git_repo_name=firewheel.git` - Defined in INSTALL.

Reviewer: That's definitely better, but I still don't love that you can only have binaries stored in one git group. Maybe I'm being too picky, so feel free to push back if you think so.

This may get ugly somewhat ugly, but what if we made some of these arguments arrays? the repo name could be in the INSTALL file like you suggest, but then there could be something like this in the firewheel config:

```yaml
ansible:
    git_servers:
        - https://github.com/sandialabs
        - https://github.com/otherrepo
        - https://gitlab.com/sandialabs
        - https://gitlab.com/otherrepo
```

Then, the system could look in each of these spots to look for the repository specified by `git_repo_name`. The biggest downside I see is that we're now checking a lot of locations, which might make it slow. But, if you only specify one repo, it's actually no slower than it would be in the current method. Thoughts?

FIREWHEEL will automatically provide the following variables to the Ansible playbooks when running:

- ``mc_name`` -- The name of the Model Component.
- ``mc_dir`` -- The full path to the model component directory.

In addition to any variables the specific tasks need, the ``vars.yml`` *should* have a ``cached_files`` key where a list of the final output files is listed.
This is because FIREWHEEL supports caching pre-computed blobs from various resources to enable offline experiment access.
The process of collecting cached files is automatically handled by FIREWHEEL and using this process is discussed in detail in :ref:`mc_install_cache`.
These cached files should be defined in ``INSTALL/vars.yml`` and the model component installation is assumed to be complete when all ``cached_files`` are present.
If no ``cached_files`` are needed, then it can be omitted from ``INSTALL/vars.yml``.

Continuing the example from above, the ``vars.yml`` would look like:

.. code-block:: yaml
  :caption: This is an example ``vars.yml`` file that ensures the final MC state.

  cached_files:
    - source: "firewheel_repo_linux/ubuntu/ubuntu/htop-1_0_2_debs.tgz"

>       destination: "{{ mc_dir }}/vm_resources/debs/htop-1_0_2_debs.tgz"

Reviewer: This example is somewhat confusing. It's supposedly continuing the example from above, but the example above is pulling from a URL. This one seems to be pulling from a git repo. 

What is the cached_files object supposed to look like when downloading from a URL? Specifically, what is the source supposed to be?

Also, adding some of this in feels a bit like duplicating the results from tasks.yml. Is that not true for some reason?

Author: I think the language of `cached_files` is still a bit nebulous. Generally, the `cached_files` source has be the same regardless of the caching method used. That's because the MC designer sets this source of the file. Therefore, the end user has to put the file in that location for their given cache. That's why standardizing the path of the files is critical. All files should follow the convention: `<package name>/<path to MC>/file`. Then the end-user would set up the cache like:

*git* = `my_cache_repo1/<package name>/<path to MC>/file`

*url* = `http://localhost:8000/<package name>/<path to MC>/file`

*S3* = `https://aws.amazon.com/mybucket/<package name>/<path to MC>/file`

I added the following to help clarify how this works.

```diff
-Continuing the example from above, the ``vars.yml`` would look like:
+Continuing the example from above, the end result of ``tasks.yml`` is the creation of the file ``{{ mc_dir }}/vm_resources/debs/htop-1_0_2_debs.tgz``.
+Rather than run the tasks (which assume online access), users could provide this file from a cache.
+The ``vars.yml`` file would look like:
```

I think this is one of the harder and less-well-defined concepts in this entire feature, so any thoughts on how to clarify this are greatly appreciated!

Author: Well, specifically, it is pulling down htop, putting it into a folder, and compressing that folder for the example experiment to use. In that case, the "cached file" would be the compressed folder. It definitely could be the debian package, if that is used directly by the MC. But in either case, the point of the cache is a check that all files that need to exist, do, and for any offline people, they can replicate that. That is part of the odd thing about cached_files, is that they sort of only exist for the purposes of offline use, (especially their "source") but are almost unnecessary for users.

As I think about this, I have some thoughts on making this a bit more user-friendly. What if the `cache_files` don't have a `source` and that the `source` is provided by FIREWHEEL. That will help with consistency of the location. Then we can have a "help" page for setting up a cache. Then to enforce that these cache files must be defined, we can enforce the convention that the script will only pass if all files in the cache are present (and have good checksums if provided). Then MC designers will have to add them but won't have any impact on how the cache is used. That may clear up the confusion around it. We could rename if necessary. Rather than "cache" it could be `required_files` or something like that.

Reviewer: Not sure I 100% see your vision for FIREWHEEL providing the `source`, but I think I'm close enough to say I support it. And yes, I think `required_files` is much clearer in this case. That feels more intuitive for my understanding of it

      firewheel_repo_dns -- Cloned repository
      └── firewheel_repo_dns
          └── dns_objects
              └── bind9_xenial_debs.tgz

    :type: string
    :required: true

.. confval:: destination

    Where the file should be placed.
    Should include ``{{ mc_path }}`` if the file needs to be relative to the model component directory.

    :type: string
    :required: true

.. confval:: checksum_algorithm checksum_algo

    Algorithm to determine checksum of file.
    Must be supported by `ansible.builtin.stat <https://docs.ansible.com/ansible/latest/collections/ansible/builtin/stat_module.html#parameter-checksum_algorithm>`_ (e.g, ``"sha1"``, ``"sha256"``, etc.).

    :type: string
    :required: false

.. confval:: checksum

    The hash of the file.

    :type: string
    :required: false

.. _mc_install_cache:

***********
Cache Types
***********

Collecting and retrieving files from a cache is automatically supported in Ansible playbooks without MC designer intervention.
Currently, FIREWHEEL supports caching files in a file server, git repository, or in an Amazon S3 data store.

> If the user sets the necessary settings in the :ref:`firewheel_configuration` for the described types below, than FIREWHEEL will automatically check those locations for the cached file.

Reviewer: ```suggestion
If the user sets the necessary settings in the :ref:`firewheel_configuration` for the described types below, then FIREWHEEL will automatically check those locations for the cached file.
```

############################

Some Model Components may require additional Python packages to be installed within FIREWHEEL's virtual environment or for data to be downloaded.

> In this case, the Model Component can have an ``INSTALL`` directory, which contains a valid `Ansible Playbook <https://docs.ansible.com/ansible/latest/playbook_guide/playbooks_intro.html>`_ (recommend method).

Reviewer: ```suggestion
In this case, the Model Component can have an ``INSTALL`` directory, which contains a valid `Ansible Playbook <https://docs.ansible.com/ansible/latest/playbook_guide/playbooks_intro.html>`_ (recommended method).
```

> Alternatively, ``INSTALL`` can be an executable script (as denoted by a `shebang <https://en.wikipedia.org/wiki/Shebang_(Unix)>`_ line), though this is not recommend and support will be removed in a future releases.

Reviewer: ```suggestion
Alternatively, ``INSTALL`` can be an executable script (as denoted by a `shebang <https://en.wikipedia.org/wiki/Shebang_(Unix)>`_ line), though this is not recommended and support will be removed in a future releases.
```

When users use the :ref:`helper_mc_generate` Helper, a new INSTALL directory is created with sample ``tasks.yml`` and ``vars.yml`` automatically included.

> When a repository is installed via the :ref:`helper_repository_install` Helper, users have the option to can automatically run each MCs INSTALL script using the ``-s`` flag (see :ref:`helper_repository_install` for more details).

Reviewer: ```suggestion
When a repository is installed via the :ref:`helper_repository_install` Helper, users have the option to automatically run each MC's INSTALL script using the ``-s`` flag (see :ref:`helper_repository_install` for more details).
```

The ``vars.yml`` file would look like:

.. code-block:: yaml
  :caption: This is an example ``vars.yml`` file that ensures the final MC state.

  required_files:
    - destination: "{{ mc_dir }}/vm_resources/debs/htop-1_0_2_debs.tgz"

The full definition for ``required_files`` is:

.. confval:: destination

    Where the file should be placed.
    Should include ``{{ mc_dir }}`` if the file needs to be relative to the model component directory.

    :type: string
    :required: true

.. confval:: source

    Where the file should be located **within** the cache.
    This should not be set by MC creators, as it defaults to ``{{ mc_name }}/file``.
    However, it is available to be modified by end-users if desired.

    :type: string
    :required: false
    :default: ``{{ mc_name }}/file``

.. confval:: checksum_algorithm

    Algorithm to determine checksum of file.
    Must be supported by `ansible.builtin.stat <https://docs.ansible.com/ansible/latest/collections/ansible/builtin/stat_module.html#parameter-checksum_algorithm>`_ (e.g, ``"sha1"``, ``"sha256"``, etc.).

    :type: string
    :required: false

.. confval:: checksum

    The hash of the file.

    :type: string

>     :required: false

Reviewer: So one issue we should probably consider is that tar archives do not get consistently hashed. So, specifying a checksum in `required_files` will work if a cached file is pulled, but not if it gets recreated by pulling the files from the regular internet. 

It looks like some fancy arguments to tar can make it consistently hashed, such as the following:

```bash
tar --mtime='2025-01-01' --sort=name --no-xattrs -cf - file1 file2 | gzip --no-name > test1.tgz
```

If the above is what we want to be the "recommended" approach, despite it being ugly, we should probably have an example of doing this with ansible. In the old install files, we made sure to check hashes of the un-compressed files before compressing, but this case is slightly different. Definitely open to better alternatives though

Author: I added some documentation to note these issues.

Users setting up a cache should place cached files using the path: ``{{ mc_name }}/{{ item.destination | basename }}``.
From the example above, the default ``source`` path would be ``linux.ubuntu/htop-1_0_2_debs.tgz``, where ``linux.ubuntu`` is the name of the associated model component.
Users can optionally modify this path by setting the :confval:`source` within the model component variables file.

Git Cache
=========
If users plan to use a git server for the Model Component cache, they can specify the following options in the :ref:`firewheel_configuration` under the ``ansible`` key.

An example of this configuration is shown below:

.. code-block:: yaml
  :caption: An example of an Ansible git server portion of the :ref:`firewheel_configuration`.

  ansible:
    git_servers:
      - server_url: "https://github.com"
        repositories:
          - path: "firewheel/mc_repo1"
          - path: "firewheel/mc_repo2"
            branch: "develop"
      - server_url: "ssh://git@gitlab.com"
        repositories:
          - path: "emulytics/firewheel/mc_repo3"
            branch: "feature-branch"
      - server_url: "https://user:ACCESS-TOKEN@github.com/"
        repositories:

>           - path: "firewheel/mc_repo4"

Reviewer: Okay I think I found a place we'll really have to optimize. It looks like EACH git repository gets cloned for EACH model component install that happens. Plus, each time it's cloned, it gets subsequently cleaned up. That means that if we're attempting to download binaries for 5 model components (pretty normal), one of which is linux, we have to clone the linux binary repo 5 times. This is going to be painfully slow and isn't necessary. So, a couple options:

1. Check to see if the required_files is satisfied after cloning each repository. That helps somewhat, but repos are still needlessly cloned frequently. We could put the biggest repos at the end so this doesn't happen as much, but still not great
2. Don't clean up cached repositories after they're cloned and don't re-clone if it's already cached. But does this create any implications for old versions of repositories stored in the cache? 
3. Hear me out... what if we don't do an LFS pull automatically? We do a non-lfs pull, check to see if the files we need are there, and then manually pull them if they exist. 

I'm open to other options, but I think as it is right now is going to be really difficult

Author: This is a great suggestion and I went with option 3. I ran a test with a non-optimal git cache configuration (where large repos without any of the required files were pulled first) and it took 477.77s while the new approach only took 187.11s, a significant time savings (290.66 seconds)! The new approach and updated documentation have now been pushed.
